torch/export/_trace.py

```diff
@@ -225,6 +247,104 @@ def convert_to_fake(x):
     return fake_args, fake_kwargs, fake_params_buffers, fake_mode
 
 
+def _skip_autograd_kernel(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    with torch._C._AutoDispatchBelowAutograd():
+        return kernel(*args, **kwargs)
+
+
+def _register_cia_to_meta(*args, **kwargs):
+    kernel = kwargs["kernel"]
+    del kwargs["kernel"]
+    if torch._C._dispatch_has_kernel_for_dispatch_key(
+        kernel.name(), torch._C.DispatchKey.CompositeImplicitAutograd
+    ):
+        return kernel._op_dk(
+            torch._C.DispatchKey.CompositeImplicitAutograd, *args, **kwargs
+        )
+    else:
+        return NotImplemented
+
+
+@contextmanager
+def override_composite_implicit_decomp(ops_to_preserve):
+    # This function overrides CompositeImplicitAutograd decomp for
+    # functional composite ops. Ideally we want to not-decompose
+    # ALL composite ops but today's C++ functinalization relies on
+    # the fact that it is working with the opset after decomp is run.
+    # Hence we can only do it for functional ops. One caveat is that
+    # there are some composite ops that lie about their schema (claimed to be
+    # functional but not really aka dropout), for these cases, we just decompose.
+    saved_tables = {}
+    patched_ops = set()
+    for op_overload in ops_to_preserve:
+
+        def can_preserve(op_overload):
```

Comment:
this function probably deserves a long comment explaining how we decide when ops can be preserved in export.